File: web/src/gf_apps/gf_crawl_lib/ts/stats/gf_crawl_stats__errors.ts
/*
GloFlow media management/publishing system
Copyright (C) 2019 Ivan Trajkovic

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

///<reference path="../../../../d/jquery.d.ts" />

namespace gf_crawl_stats__errors {

//---------------------------------------------------
export function init(p_log_fun) {
	p_log_fun("FUN_ENTER", "gf_crawl_stats__errors.init()");

	$('#view_crawl_errors_btn').on('click', (p_stats_map)=>{

		gf_stats.http__stats_query('errors',
			(p_stats_map)=>{
				const errors_stats = view__errors_stats(p_stats_map,p_log_fun);

				$('#crawl_errors #errors').append(errors_stats);
			},
			()=>{},
			p_log_fun);
	});
}
//---------------------------------------------------
function view__errors_stats(p_stats_map, p_log_fun) {
	p_log_fun("FUN_ENTER", "gf_crawl_stats__errors.view__errors_stats()");

	const stats = $(`<div id="errors_stats"></div>`);

	const stat_errors_lst = p_stats_map['stat_errors_lst'];
	for (var crawler_errors_map of stat_errors_lst) {

		const crawler_name_str = crawler_errors_map['crawler_name_str'];
		const errors_types_lst = crawler_errors_map['errors_types_lst'];

		const crawler_errors = $(`<div class="crawler_errors">
				<div class='crawler_name'>`+crawler_name_str+`</div>
			</div>`);

		$(stats).append(crawler_errors);

		for (var errors_type_map of errors_types_lst) {
			const error_type_str = errors_type_map['type_str'];
			const count_int      = errors_type_map['count_int'];
			const urls_lst       = errors_type_map['urls_lst'];

			const errors_type = $(`
				<div class='errors_type'>
					<div><span class='label'>type:</span><span>`+error_type_str+`</span></div>
					<div><span class='label'>count:</span><span>`+count_int+`</span></div>
					<div class='urls_lst'></div>
				</div>`);
			
			for (var url_str of urls_lst) {
				$(errors_type).find('.urls_lst').append($('<div class="url"><a target="_blank" href="'+url_str+'">'+url_str+'</a></div>'))
			}

			$(crawler_errors).append(errors_type);
		}        
	}

	return stats;
}
//---------------------------------------------------
}

File: go/gf_apps/gf_crawl_lib/gf_crawl_stats/gf_crawl_stats__errors.go
/*
GloFlow application and media management/publishing platform
Copyright (C) 2019 Ivan Trajkovic

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

package gf_crawl_stats

import (
	"context"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/bson"
	// "github.com/globalsign/mgo/bson"
	"github.com/gloflow/gloflow/go/gf_core"
)

//-------------------------------------------------
type Gf_stat__errors struct {
	Crawler_name_str string                `bson:"_id"              json:"crawler_name_str"`
	Errors_types_lst []Gf_stat__error_type `bson:"errors_types_lst" json:"errors_types_lst"`
}

type Gf_stat__error_type struct {
	Type_str  string   `bson:"type_str"  json:"type_str"`
	Count_int int      `bson:"count_int" json:"count_int"`
	Urls_lst  []string `bson:"urls_lst"  json:"urls_lst"`
}

//-------------------------------------------------
func stats__errors(p_runtime_sys *gf_core.Runtime_sys) (map[string]interface{}, *gf_core.Gf_error) {
	p_runtime_sys.Log_fun("FUN_ENTER","gf_crawl_stats__errors.stats__errors()")


	ctx := context.Background()
	pipeline := mongo.Pipeline{
		{
			{"$match", bson.D{{"t", "crawler_error"}}},
		},
		{
			{"$sort", bson.D{{"creation_unix_time_f", -1}}},
		},
		{
			{"$group", bson.D{
				{"_id",                     bson.M{"type_str": "$type_str", "crawler_name_str": "$crawler_name_str"}},
				{"count_int",               bson.M{"$sum": 1}},
				{"urls_lst",                bson.M{"$push": "$url_str"}},
				{"creation_unix_times_lst", bson.M{"$push": "$creation_unix_time_f"}},
			}},
		},
		{
			{"$group", bson.D{
				{"_id", "$_id.crawler_name_str"},
				{"errors_types_lst", bson.M{"$push": bson.M{
					"type_str":                "$_id.type_str",
					"count_int":               "$count_int",
					"urls_lst":                "$urls_lst",
					"creation_unix_times_lst": "$creation_unix_times_lst",
				}}},
			}},
		},
	}

	/*pipe := p_runtime_sys.Mongo_db.Collection("gf_crawl").Pipe([]bson.M{
		bson.M{"$match":bson.M{
				"t": "crawler_error",
			},
		},

		bson.M{"$sort": bson.M{
				"creation_unix_time_f": -1,
			},
		},

		bson.M{"$group":bson.M{
				"_id":                     bson.M{"type_str": "$type_str", "crawler_name_str": "$crawler_name_str",},
				"count_int":               bson.M{"$sum": 1},
				"urls_lst":                bson.M{"$push": "$url_str"},
				"creation_unix_times_lst": bson.M{"$push": "$creation_unix_time_f"},
			},
		},

		bson.M{"$group":bson.M{
				"_id":              "$_id.crawler_name_str",
				"errors_types_lst": bson.M{"$push": bson.M{
							"type_str":                "$_id.type_str",
							"count_int":               "$count_int",
							"urls_lst":                "$urls_lst",
							"creation_unix_times_lst": "$creation_unix_times_lst",
						},
					},
			},
		},
	})*/


	cursor, err := p_runtime_sys.Mongo_coll.Aggregate(ctx, pipeline)
	if err != nil {

		gf_err := gf_core.Mongo__handle_error("failed to run an aggregation pipeline to count/get_info of crawler_error's by crawler_name",
			"mongodb_aggregation_error",
			map[string]interface{}{},
			err, "gf_crawl_stats", p_runtime_sys)
		return nil, gf_err
	}
	defer cursor.Close(ctx)


	/*results_lst := []Gf_stat__errors{}
	err         := pipe.AllowDiskUse().All(&results_lst)

	if err != nil {
		gf_err := gf_core.Mongo__handle_error("failed to run an aggregation pipeline to count/get_info of crawler_error's by crawler_name",
			"mongodb_aggregation_error",
			nil, err, "gf_crawl_stats", p_runtime_sys)
		return nil, gf_err
	}*/
	
	results_lst := []Gf_stat__errors{}
	for cursor.Next(ctx) {

		var r Gf_stat__errors
		err := cursor.Decode(&r)
		if err != nil {
			gf_err := gf_core.Mongo__handle_error("failed to run an aggregation pipeline to count/get_info of crawler_error's by crawler_name",
				"mongodb_cursor_decode",
				map[string]interface{}{},
				err, "gf_crawl_stats", p_runtime_sys)
			return nil, gf_err
		}
	
		results_lst = append(results_lst, r)
	}

	data_map := map[string]interface{}{
		"errors_lst": results_lst,
	}
	return data_map, nil
}